core/services/p2p/peer.go

```diff
@@ -0,0 +1,250 @@
+	DeltaReconcile time.Duration
+
+	// Dial attempts will be at least DeltaDial apart.
+	DeltaDial time.Duration
+
+	DiscovererDatabase nettypes.DiscovererDatabase
+
+	IncomingMessageBufferSize int
+	OutgoingMessageBufferSize int
+	MaxMessageLenBytes        int
+	MessageRateLimiter        ragep2p.TokenBucketParams
+	BytesRateLimiter          ragep2p.TokenBucketParams
+
+	MetricsRegisterer prometheus.Registerer
+}
+
+type peer struct {
+	streams     map[ragetypes.PeerID]*ragep2p.Stream
+	cfg         PeerConfig
+	isBootstrap bool
+	host        *ragep2p.Host
+	discoverer  *ragedisco.Ragep2pDiscoverer
+	myID        ragetypes.PeerID
+	recvCh      chan Message
+
+	stopCh services.StopChan
+	wg     sync.WaitGroup
+	lggr   logger.Logger
+}
+
+func NewPeer(cfg PeerConfig, lggr logger.Logger) (Peer, error) {
+	peerID, err := ragetypes.PeerIDFromPrivateKey(cfg.PrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting v2 peer ID from private key: %w", err)
+	}
+	isBootstrap := false
+	for _, info := range cfg.Bootstrappers {
+		if info.ID == peerID {
+			isBootstrap = true
+			break
+		}
+	}
+
+	announceAddresses := cfg.AnnounceAddresses
+	if len(cfg.AnnounceAddresses) == 0 {
+		announceAddresses = cfg.ListenAddresses
+	}
+
+	discoverer := ragedisco.NewRagep2pDiscoverer(cfg.DeltaReconcile, announceAddresses, cfg.DiscovererDatabase, cfg.MetricsRegisterer)
+	commonLggr := commonlogger.NewOCRWrapper(lggr, true, func(string) {})
+
+	host, err := ragep2p.NewHost(
+		ragep2p.HostConfig{DurationBetweenDials: cfg.DeltaDial},
+		cfg.PrivateKey,
+		cfg.ListenAddresses,
+		discoverer,
+		commonLggr,
+		cfg.MetricsRegisterer,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct ragep2p host: %w", err)
+	}
+
+	return &peer{
+		streams:     make(map[ragetypes.PeerID]*ragep2p.Stream),
+		cfg:         cfg,
+		isBootstrap: isBootstrap,
+		host:        host,
+		discoverer:  discoverer,
+		myID:        peerID,
+		recvCh:      make(chan Message, cfg.IncomingMessageBufferSize),
+		stopCh:      make(services.StopChan),
+		lggr:        lggr.Named("P2PPeer"),
+	}, nil
+}
+
+func (p *peer) UpdatePeerAddresses(peerIDs []ragetypes.PeerID) error {
+	p.lggr.Infow("updating peer addresses", "peerIDs", peerIDs)
+	if !p.isBootstrap {
+		if err := p.recreateStreams(peerIDs); err != nil {
+			return err
+		}
+	}
+
+	if err := p.discoverer.RemoveGroup(defaultGroupID); err != nil {
+		p.lggr.Warnw("failed to remove old group", "groupID", defaultGroupID)
+	}
+	if err := p.discoverer.AddGroup(defaultGroupID, peerIDs, p.cfg.Bootstrappers); err != nil {
+		p.lggr.Warnw("failed to add group", "groupID", defaultGroupID)
+		return err
+	}
+	return nil
+}
+
+func (p *peer) recreateStreams(peerIDs []ragetypes.PeerID) error {
+	p.lggr.Infow("recreateStreams", "peerIDs", peerIDs)
+	peerIDSet := make(map[ragetypes.PeerID]struct{})
+	for _, pid := range peerIDs {
+		pid := pid
+		if pid == p.myID { // don't create a self-stream
+			continue
```

Comment:
Is it common in golang to use continues in for loops? I was always told to avoid using them in most languages because they act a bit like jumps and can make it a little harder to understand control flow in loops. It's pretty easy to avoid using them:

```
if pid != p.myID {
   .... rest of the for loop logic here
}
```

Reply:
You can make arguments either way. I remember reading an opinion that the more nested some logic is the harder it is to understand. I don't know of any Go-specific recommendations, maybe @jmank88 knows.

Reply:
I haven't seen much objection to `continue` in go, but there is definitely a tendency towards reducing nesting/indentation, so yeah maybe that alleviates this concern.

I suppose that there is an argument that using a continue in a for like this is actually analogous to the go proverb "return early, return often", which is meant to reduce indentation :shrug: